organisations/models.py

```diff
@@ -37,29 +37,51 @@ def user_can_view(self, user):
             return self.none()
 
     def retention_period_exceeded(self):
+        """
+        Return queryset of Person objects whose personal data
+        retention period has exceeded.
+
+        This excludes:
+        - Persons with a user account
+        - Persons with any valid (recent) enrolments
+        """
         earliest_valid_timestamp = timezone.now() - relativedelta(
             months=settings.PERSONAL_DATA_RETENTION_PERIOD_MONTHS
         )
 
-        all_enrolments_too_old = Q(max_enrolment_end_time=None) | Q(
-            max_enrolment_end_time__lt=earliest_valid_timestamp
-        )
-        all_study_groups_too_old = Q(max_studygroup_end_time=None) | Q(
-            max_studygroup_end_time__lt=earliest_valid_timestamp
-        )
-
-        return (
-            # At least for now returns only unauthenticated Person's
-            self.filter(user=None)
-            .annotate(
-                max_enrolment_end_time=Max("enrolment__occurrence__end_time"),
-                max_studygroup_end_time=Max(
-                    "studygroup__enrolments__occurrence__end_time"
-                ),
+        # Find persons with any valid enrolments or
+        # event queue enrolments (as person or study group contact)
+        valid_enrolment_persons = self.filter(
+            Q(enrolment__occurrence__end_time__gt=earliest_valid_timestamp)
+            | Q(
+                studygroup__enrolments__occurrence__end_time__gt=earliest_valid_timestamp
+            )
+            | Q(
+                eventqueueenrolment__p_event__occurrences__end_time__gt=earliest_valid_timestamp
+            )
+        ).distinct()
+
+        # Return persons who:
+        # 1. Don't have a user account
+        # 2. Don't have any valid enrolments (either as person or study group contact)
+        # 3. Don't have any valid event queue enrolments
+        old_enrolment_filter = (
+            Q(enrolment__occurrence__end_time__lte=earliest_valid_timestamp)
+            | Q(enrolment__occurrence__end_time__isnull=True)
+            | Q(
+                studygroup__enrolments__occurrence__end_time__lte=earliest_valid_timestamp
             )
-            .filter(all_enrolments_too_old & all_study_groups_too_old)
+            | Q(studygroup__enrolments__occurrence__end_time__isnull=True)
+            | Q(
+                eventqueueenrolment__p_event__occurrences__end_time__lte=earliest_valid_timestamp
+            )
+            | Q(eventqueueenrolment__p_event__occurrences__end_time__isnull=True)
         )
 
+        persons = self.filter(user__isnull=True).filter(old_enrolment_filter)
+        persons = persons.exclude(id__in=valid_enrolment_persons.values("id"))
+        return persons.distinct()
```

Comment:
Given 

```python
        # Return persons who:
        # 1. Don't have a user account
        # 2. Don't have any valid enrolments (either as person or study group contact)
        # 3. Don't have any valid event queue enrolments
```
Unless I'm missing something, the Q logic got unnecessarily complicated here? Tests do pass with something like:

```python
return (
    self
    .exclude(user__isnull=False)
    .exclude(enrolment__occurrence__end_time__gt=earliest_valid_timestamp)
    .exclude(studygroup__enrolments__occurrence__end_time__gt=earliest_valid_timestamp)
    .exclude(eventqueueenrolment__p_event__occurrences__end_time__gt=earliest_valid_timestamp)
)
```

Reply:
plus

```python
.exclude(studygroup__queued_enrolments__p_event__occurrences__end_time__gt=earliest_valid_timestamp)
```

?